crates/burn-autodiff/src/runtime/memory_management.rs

```diff
@@ -115,81 +107,98 @@ impl GraphMemoryManagement {
+
+            let node_status = self
+                .statuses
+                .get(&node_id)
+                .expect("All nodes should have received a status at this point");
+
+            // A node is useful if it is not tagged as unavailable, and if
+            // it is directly referenced or needed by the node that called it
+            if let NodeMemoryStatus::Unknown = node_status {
+                if self.is_referenced(node_id) {
+                    self.statuses.insert(node_id, NodeMemoryStatus::Useful);
+                    visited_as_useful.insert(node_id);
+                } else if let Some(caller_id) = caller_id {
+                    let caller_status = self
+                        .statuses
+                        .get(&caller_id)
+                        .expect("Caller should have status");
+                    if let NodeMemoryStatus::Useful = caller_status {
+                        self.statuses.insert(node_id, NodeMemoryStatus::Useful);
+                        visited_as_useful.insert(node_id);
+                    }
                 }
             }
-            Mode::Explore => {
-                let node_status = self
-                    .statuses
-                    .get(&node_id)
-                    .expect("All nodes should have received a status at this point")
-                    .clone();
-
-                match node_status {
-                    NodeMemoryStatus::Useful => {
-                        // Nothing to do, was already tagged through some other path
-                    }
-                    NodeMemoryStatus::Unavailable => {
-                        // Even if this node is unavailable, it is still possible that an ancestor is useful if referenced
-                        for parent in parents {
-                            self.useful_propagation(parent, Mode::Explore);
-                        }
-                    }
-                    NodeMemoryStatus::Unknown => {
-                        // If this node is referenced and not unavailable,
-                        // then it is useful and we must retain all ancestors
-
-                        let mut mode = Mode::Explore;
-                        if self.is_referenced(node_id) {
-                            self.statuses.insert(node_id, NodeMemoryStatus::Useful);
-                            mode = Mode::TagAsUseful;
-                        }
 
-                        for parent in parents {
-                            self.useful_propagation(parent, mode.clone());
-                        }
-                    }
+            for parent in self
+                .nodes
+                .get(&node_id)
+                .cloned()
+                .unwrap_or_default()
+                .into_iter()
+            {
+                if !visited_as_useful.contains(&parent)
+                    && (Some(&NodeMemoryStatus::Useful) == self.statuses.get(&node_id)
+                        || !visited_as_unknown.contains(&parent))
```

Comment:
I would refactor that a bit, very hard to read:

```rust
if visited_as_useful.contains(&parents); {
    continue;
}

let is_useful = Some(&NodeMemoryStatus::Useful) == self.statuses.get(&node_id);

if is_useful || !visited_as_unknown.contains(&parents) {
   to_visit.push((parent, Some(node_id)))
}
```

And I actually think there is still a performance problem and even a bug. I think it should be:

```rust
if visited_as_useful.contains(&parents); {
    continue;
}

let is_useful = Some(&NodeMemoryStatus::Useful) == self.statuses.get(&node_id);

if is_useful {
   to_visit.push((parent, Some(node_id)))
}
```

Since the vector `to_visit` is already filled with all nodes, you don't need to keep track of nodes that were visited as unknown.

Reply:
I rewrote the algorithm, it's more efficient and elegant